eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelayer.scala

```diff
@@ -95,26 +100,40 @@ class NodeRelayer(nodeParams: NodeParams, relayer: ActorRef, router: ActorRef, c
           case None =>
             log.info(s"relaying trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv} htlcCount=${parts.length})")
             val paymentId = relay(paymentHash, upstream, nextPayload, nextPacket)
-            context become main(pendingIncoming - paymentHash, pendingOutgoing + (paymentId -> PendingResult(upstream, nextPayload)))
+            context become main(pendingIncoming - paymentHash, pendingOutgoing + (paymentHash -> PendingResult(upstream, nextPayload, paymentId, settled = false)))
         }
-      case None => throw new RuntimeException(s"could not find pending incoming payment (paymentHash=$paymentHash)")
+      case None => log.error("could not find pending incoming payment: payment will not be relayed: please investigate")
     }
 
+    case Relayer.ForwardFulfill(fulfill, Origin.TrampolineRelayed(_, Some(paymentSender)), _) =>
+      paymentSender ! fulfill
+      val paymentHash = Crypto.sha256(fulfill.paymentPreimage)
+      pendingOutgoing.get(paymentHash).foreach(p => if (!p.settled) {
+        // We want to fulfill upstream as soon as we receive the preimage (even if not all HTLCs have fulfilled downstream).
+        log.debug("trampoline payment successfully relayed")
+        fulfillPayment(p.upstream, fulfill.paymentPreimage)
+        context become main(pendingIncoming, pendingOutgoing + (paymentHash -> p.copy(settled = true)))
+      })
+
     case PaymentSent(id, paymentHash, paymentPreimage, _, _, parts) =>
-      log.debug("trampoline payment successfully relayed")
-      pendingOutgoing.get(id).foreach {
-        case PendingResult(upstream, _) =>
-          fulfillPayment(upstream, paymentPreimage)
-          val incoming = upstream.adds.map(add => PaymentRelayed.Part(add.amountMsat, add.channelId))
-          val outgoing = parts.map(part => PaymentRelayed.Part(part.amountWithFees, part.toChannelId))
-          context.system.eventStream.publish(TrampolinePaymentRelayed(paymentHash, incoming, outgoing))
-      }
-      context become main(pendingIncoming, pendingOutgoing - id)
-
-    case PaymentFailed(id, _, failures, _) =>
-      log.debug("trampoline payment failed")
-      pendingOutgoing.get(id).foreach { case PendingResult(upstream, nextPayload) => rejectPayment(upstream, translateError(failures, nextPayload.outgoingNodeId)) }
-      context become main(pendingIncoming, pendingOutgoing - id)
+      // We may have already fulfilled upstream, but we can now emit an accurate relayed event and clean-up resources.
```

Comment:
I would have factored the whole fulfillment process and done it the same way in both handlers, because a payment is really sent when we forward the preimage upstream. No real reasons to distinguish I think.